diff --git a/dist/installHooks.js b/dist/installHooks.js
index f1d3626df21b417c799eac2bbcc6ded06ebe90af..0a98817fb7abe6c4550458e9a4a5ed5f3a1e0f39 100644
--- a/dist/installHooks.js
+++ b/dist/installHooks.js
@@ -105,27 +105,13 @@ function installHooks(db, encryptionOptions, keyPromise, performEncryption, perf
                                 if (!cursor) {
                                     return cursor;
                                 }
-                                return Object.create(cursor, {
-                                    continue: {
-                                        get() {
-                                            return cursor.continue;
-                                        },
-                                    },
-                                    continuePrimaryKey: {
-                                        get() {
-                                            return cursor.continuePrimaryKey;
-                                        },
-                                    },
-                                    key: {
-                                        get() {
-                                            return cursor.key;
-                                        },
-                                    },
-                                    value: {
-                                        get() {
-                                            return decrypt(cursor.value);
-                                        },
-                                    },
+                                return new Proxy(cursor, {
+                                    get(target, prop) {
+                                        if (prop === 'value') {
+                                            return decrypt(target.value)
+                                        }
+                                        return target[prop]
+                                    }
                                 });
                             });
                         },
