patch-package
--- a/node_modules/dexie/dist/dexie.es.js
+++ b/node_modules/dexie/dist/dexie.es.js
@@ -4413,9 +4413,18 @@ props(Dexie, {
     // For node.js, you need to require indexeddb-js or similar and then set these deps.
     //
     dependencies: {
-        // Required:
-        indexedDB: _global.indexedDB || _global.mozIndexedDB || _global.webkitIndexedDB || _global.msIndexedDB,
-        IDBKeyRange: _global.IDBKeyRange || _global.webkitIDBKeyRange
+      // Required:
+      get indexedDB() {
+        try {
+          return _global.indexedDB || _global.mozIndexedDB || _global.webkitIndexedDB || _global.msIndexedDB;
+        } catch(e) {}
+
+      },
+      get IDBKeyRange() {
+        try {
+          return _global.IDBKeyRange || _global.webkitIDBKeyRange;
+        } catch(e) {}
+      }
     },
     // API Version Number: Type Number, make sure to always set a version number that can be comparable correctly. Example: 0.9, 0.91, 0.92, 1.0, 1.01, 1.1, 1.2, 1.21, etc.
     semVer: DEXIE_VERSION,
